cirq/experiments/fidelity_estimation.py

```diff
@@ -0,0 +1,51 @@
+"""Estimate fidelity of large random quantum circuit from observed bitstrings.
+
+Fidelity estimator defined here is used in cross-entropy benchmarking and
+works under the assumption that the evaluated circuit is sufficiently
+random, see https://arxiv.org/abs/1608.00263.
+"""
+
+from typing import cast, List, Sequence, Tuple
+
+import numpy as np
+
+from cirq.circuits import Circuit
+from cirq.ops import Qid
+from cirq.sim import Simulator, WaveFunctionTrialResult
+
+
+def estimate_circuit_fidelity(circuit: Circuit, qubit_order: Sequence[Qid],
+                              bitstrings: Sequence[int]) -> float:
+    """Computes fidelity estimate from one circuit using linear XEB estimator.
+
+    Args:
+        circuit: Random quantum circuit which has been executed on quantum
+            processor under test
+        qubit_order: Qubit order used to construct bitstrings from measurements
+        bitstrings: Results of terminal all-qubit measurements performed after
+            each circuit execution
+    Returns:
+        Estimate of circuit fidelity.
+    Raises:
+        ValueError: Circuit is inconsistent with qubit order or one of the
+            bitstrings is inconsistent with the number of qubits.
+    """
+    dim = 2**len(qubit_order)
+
+    if set(qubit_order) != circuit.all_qubits():
+        raise ValueError(
+            f'Inconsistent qubits: circuit has {circuit.all_qubits()}, '
+            f'qubit order is {qubit_order}')
+    for bitstring in bitstrings:
+        if bitstring < 0 or dim <= bitstring:
+            raise ValueError(
+                f'Bitstring {bitstring} could not have been observed '
+                f'on {len(qubit_order)} qubits.')
+
+    simulator = Simulator()
+    result = cast(WaveFunctionTrialResult,
+                  simulator.simulate(circuit, qubit_order=qubit_order))
```

Comment:
If the bitstrings are in a numpy array as I suggest above, then you can use `simulator.compute_amplitudes`.

Reply:
Another comment suggested `cirq.final_wavefunction()` which plays well with the way we specify bitstrings here. That said, `compute_amplitudes()` does have an advantage since in future it may be able to compute a proper subset of amplitudes which would allow this code to work with (slightly) larger number of qubits. Then again, for convenience (compact output) and compatibility with other code I'd like to represent measurements as int arrays not as 2-D bool arrays...

What do y'all think about changing the interface (2D bool -> 1D int for measurements)? @kevinsung @Strilanc

Reply:
I think the bitstrings input to `compute_amplitudes` (as well as your `estimate_circuit_fidelity`) should be of the same format as the bitstrings in `TrialResult.measurements`. Doesn't that better satisfy your "compatibility with other code" criterion?

Reply:
I see now that the way you have it is compatible with `TrialResult.data`, so I'm fine with it actually. I agree that we should change the interface of `compute_amplitudes` to match with this.

Reply:
FTR: #2151.